Initial support for loongarch64-unknown-linux-gnu (#96971)
rust-lang/rust: src/doc/rustc/src/platform-support/loongarch-linux.md

# loongarch\*-unknown-linux-\*

**Tier: 3**

[LoongArch] is a new RISC ISA developed by Loongson Technology, based in China.

[LoongArch]: https://loongson.github.io/LoongArch-Documentation/README-EN.html

> The target name follow this format: `<machine>-<vendor>-<os><fabi_suffix>, where `<machine>` specifies the CPU family/model, `<vendor>` specifies the vendor and `<os>` the operating system name.

Reviewer: Self note: rewrite this document in 2 days.

Reviewer: Hmm. I'll probably have to hurry a bit...

Reviewer 2: Should we try ChatGPT here? :)

Reviewer: I'll be revising this doc later in a separate PR (at least one change to the tier is definitely going to happen), so consider my previous comment to no longer block this PR.

|           lp64d/base             |   glibc     |  linux   | loongarch64-unknown-linux-gnuf64 |
|           lp64f/base             |   glibc     |  linux   | loongarch64-unknown-linux-gnuf32 |
|           lp64s/base             |   glibc     |  linux   | loongarch64-unknown-linux-gnusf  |
|           ilp32d/base            |   glibc     |  linux   | loongarch32-unknown-linux-gnuf64 |
|           ilp32f/base            |   glibc     |  linux   | loongarch32-unknown-linux-gnuf32 |
|           ilp32s/base            |   glibc     |  linux   | loongarch32-unknown-linux-gnusf  |
|           lp64d/base             |  musl libc  |  linux   | loongarch64-unknown-linux-muslf64|
|           lp64f/base             |  musl libc  |  linux   | loongarch64-unknown-linux-muslf32|
|           lp64s/base             |  musl libc  |  linux   | loongarch64-unknown-linux-muslsf |
|           ilp32d/base            |  musl libc  |  linux   | loongarch32-unknown-linux-muslf64|
|           ilp32f/base            |  musl libc  |  linux   | loongarch32-unknown-linux-muslf32|
|           ilp32s/base            |  musl libc  |  linux   | loongarch32-unknown-linux-muslsf |

## Target maintainers

> - zhaixiaojuan, `zhaixiaojuan@loongson.cn`, https://github.com/zhaixiaojuan

Reviewer 3: Are there any other people that should be listed here? I see a few different people in this PR 🙂

Reviewer 2: If a third-party maintainer (not affiliated with Loongson) is wanted I guess we can list @xen0n here.

I've ported Rustc to LoongArch64 independently so maybe I can be listed too, but frankly I'm not a Rust expert.

## Requirements

This target is cross-compiled.
A GNU toolchain for LoongArch target is required.  It can be downloaded from https://github.com/loongson/build-tools/releases, or built from the source code of GCC (12.1.0 or later) and Binutils (2.39 or later).

## Building the target

The target can be built by enabling it for a `rustc` build.

```toml
[build]
target = ["loongarch64-unknown-linux-gnuf64"]
```

Make sure `loongarch64-unknown-linux-gnu-gcc` can be searched from the directories specified in`$PATH`. Alternatively, you can use GNU LoongArch Toolchain by adding the following to `config.toml`:

```toml
[target.loongarch64-unknown-linux-gnuf64]
# ADJUST THIS PATH TO POINT AT YOUR TOOLCHAIN
cc = "/TOOLCHAIN_PATH/bin/loongarch64-unknown-linux-gnu-gcc"
cxx = "/TOOLCHAIN_PATH/bin/loongarch64-unknown-linux-gnu-g++"
ar = "/TOOLCHAIN_PATH/bin/loongarch64-unknown-linux-gnu-ar"
ranlib = "/TOOLCHAIN_PATH/bin/loongarch64-unknown-linux-gnu-ranlib"
linker = "/TOOLCHAIN_PATH/bin/loongarch64-unknown-linux-gnu-gcc"
```

## Cross-compilation

This target can be cross-compiled on a `x86_64-unknown-linux-gnu` host. Cross-compilation on other hosts may work but is not tested.

## Testing

> Currently there is no support to run the rustc test suite for this target.

Reviewer 3: Is there an emulator/QEMU support for this architecture? That would be good to mention in this doc.

Author: @wesleywiser  Thanks a lot, I have updated the code.

> Is there an emulator/QEMU support for this architecture? That would be good to mention in this doc.

A simple example of testing Loongarch64 architecture binaries on the x86_86 platform: [test expample](https://github.com/rust-lang/rust/commit/8e8dee62ebbf1847e69b9442bfc9207760a00a6f)
If the rust repository needs to run related tests on the loongarch64 architecture, we can also provide remote physical machines.

# loongarch\*-unknown-linux-\*

**Tier: 3**

[LoongArch] LoongArch is a RISC style ISA which is independently designed by Loongson Technology in China.

[LoongArch]: https://loongson.github.io/LoongArch-Documentation/README-EN.html

> The target name follow this format: `<machine>-<vendor>-<os><fabi_suffix><abiext_suffix>`,where `<machine>` specifies the CPU family/model, `<vendor>` specifies the vendor and `<os>` the operating system name.

Reviewer 2: Should we remove "abiext_suffix" for now as it's not use anywhere?

|           lp64s/base             |   glibc     |  linux   | loongarch64-unknown-linux-gnusf  |
|           ilp32d/base            |   glibc     |  linux   | loongarch32-unknown-linux-gnuf64 |
|           ilp32f/base            |   glibc     |  linux   | loongarch32-unknown-linux-gnuf32 |
|           ilp32s/base            |   glibc     |  linux   | loongarch32-unknown-linux-gnusf  |
|           lp64d/base             |  musl libc  |  linux   | loongarch64-unknown-linux-muslf64|
|           lp64f/base             |  musl libc  |  linux   | loongarch64-unknown-linux-muslf32|
|           lp64s/base             |  musl libc  |  linux   | loongarch64-unknown-linux-muslsf |
|           ilp32d/base            |  musl libc  |  linux   | loongarch32-unknown-linux-muslf64|
|           ilp32f/base            |  musl libc  |  linux   | loongarch32-unknown-linux-muslf32|
|           ilp32s/base            |  musl libc  |  linux   | loongarch32-unknown-linux-muslsf |

## Target maintainers

- zhaixiaojuan, `zhaixiaojuan@loongson.cn`, https://github.com/zhaixiaojuan

## Requirements

This target is cross-compiled.

> A platform-provided C compiler toolchain is required, and obtaining address of the cross toolchain [GNU LoongArch toolchain](https://github.com/loongson/build-tools/releases).

Reviewer 2: ```suggestion
A GNU toolchain for LoongArch target is required.  It can be downloaded from https://github.com/loongson/build-tools/releases, or built from the source code of GCC (12.1.0 or later) and Binutils (2.39 or later).
```

I'm not really sure about how to specify the minimal Binutils version, as 2.39 not released yet but 2.38 does not work.

## Building the target

The target can be built by enabling it for a `rustc` build.

```toml
[build]
target = ["loongarch64-unknown-linux-gnuf64"]
```

> Make sure `loongarch64-unknown-linux-gnu-gcc` is included in `$PATH`. Alternatively, you can use GNU LoongArch Toolchain by adding the following to `config.toml`:

Reviewer 2: ```suggestion
Make sure `loongarch64-unknown-linux-gnu-gcc` can be searched from the directories specified in`$PATH`. Alternatively, you can use GNU LoongArch Toolchain by adding the following to `config.toml`:
```

Reviewer: "Make sure `loongarch64-unknown-linux-gnu-gcc` is in `$PATH`." is enough. "can be searched from" is Chinglish too ;-)

Reviewer 2: I think "in `$PATH`" is some sort of ambigious (it's a directory listed in `$PATH` or a file in those directories?)

The following is copied from "man execl":

"Can be sought in the directory pathnames specified in the PATH environment variable."

```toml
[target.loongarch64-unknown-linux-gnuf64]

> cc = "loongarch64-unknown-linux-gnu-gcc"

Reviewer 2: ```suggestion
cc = "/path/to/loongarch64-unknown-linux-gnu-gcc"
```

I think it's better to use an absolute path for these examples?  With no explicit path it's same as a search in `$PATH`.

cxx = "loongarch64-unknown-linux-gnu-g++"
ar = "loongarch64-unknown-linux-gnu-ar"
ranlib = "loongarch64-unknown-linux-gnu-ranlib"
linker = "loongarch64-unknown-linux-gnu-gcc"
```

## Cross-compilation

> This target can be cross-compiled as `x86_64`, and cross-compilation on other hosts is not currently supported.

Reviewer 2: ```suggestion
This target can be cross-compiled on a `x86_64-unknown-linux-gnu` host.  Cross-compilation on other hosts may work but is not tested.
```

Reviewer: "This target can be cross-compiled. Cross-compiling from `x86_64-unknown-linux-gnu` is known to work; any other host will likely work too."

|          f64           | The base ABI use 64-bits FPRs for parameter passing.(lp64d)|
|          f32           | The base ABI uses 32-bit FPRs for parameter passing. (lp64f)|
|          sf            | The base ABI uses no FPR for parameter passing. (lp64s)     |

|`ABI type(Base ABI/ABI extension)`| `C library` | `kernel` |          `target tuple`          |
|----------------------------------|-------------|----------|----------------------------------|
|           lp64d/base             |   glibc     |  linux   | loongarch64-unknown-linux-gnu |
|           lp64f/base             |   glibc     |  linux   | loongarch64-unknown-linux-gnuf32 |
|           lp64s/base             |   glibc     |  linux   | loongarch64-unknown-linux-gnusf  |
|           lp64d/base             |  musl libc  |  linux   | loongarch64-unknown-linux-musl|
|           lp64f/base             |  musl libc  |  linux   | loongarch64-unknown-linux-muslf32|
|           lp64s/base             |  musl libc  |  linux   | loongarch64-unknown-linux-muslsf |

## Target maintainers

- zhaixiaojuan, `zhaixiaojuan@loongson.cn`, https://github.com/zhaixiaojuan

## Requirements

This target is cross-compiled.

> A GNU toolchain for LoongArch target is required.  It can be downloaded from https://github.com/loongson/build-tools/releases, or built from the source code of GCC (12.1.0 or later) and Binutils (2.39 or later).

Reviewer 2: ```suggestion
A GNU toolchain for LoongArch target is required.  It can be downloaded from https://github.com/loongson/build-tools/releases, or built from the source code of GCC (12.1.0 or later) and Binutils (2.40 or later).
```

We are setting `EF_LARCH_OBJABI_V1`, and our LLVM backend is emitting non-stack-based relocation.  So ld-2.39 will not work.

# loongarch\*-unknown-linux-\*

**Tier: 3**

[LoongArch] is a new RISC ISA developed by Loongson Technology Corporation Limited.

[LoongArch]: https://loongson.github.io/LoongArch-Documentation/README-EN.html

> The target name follow this format: `<machine>-<vendor>-<os><fabi_suffix>, where `<machine>` specifies the CPU family/model, `<vendor>` specifies the vendor and `<os>` the operating system name.

Reviewer 4: ```suggestion
The target name follow this format: `<machine>-<vendor>-<os><fabi_suffix>`, where `<machine>` specifies the CPU family/model, `<vendor>` specifies the vendor and `<os>` the operating system name.
```

Reviewer 5: Oops, Is it better to fix this PR or a new one? thanks

Reviewer 4: A new PR works better I think given that this PR is already included in a rollup.
